src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -517,12 +545,53 @@ private static bool IsFlagSet(Interop.EventStream.FSEventStreamEventFlags flags,
                 return (value & flags) == value;
             }
 
-            private static bool DoesItemExist(string path, bool isFile)
+            private static bool DoesItemExist(ReadOnlySpan<char> path, bool isFile)
             {
-                if (isFile)
-                    return File.Exists(path);
-                else
-                    return Directory.Exists(path);
+                Interop.ErrorInfo ignored;
+
+                try
+                {
+                    if (path.IsEmpty || path.Length == 0)
+                        return false;
+
+                    if(!isFile)
+                    {
+                        return Exists(path, Interop.Sys.FileTypes.S_IFDIR, out ignored);
+                    }
+
+                    return path.Length > 0 && PathInternal.IsDirectorySeparator(path[path.Length - 1])
+                        ? false
+                        : Exists(path, Interop.Sys.FileTypes.S_IFREG, out ignored);
+                }
+                catch (ArgumentException) { }
+                catch (IOException) { }
+                catch (UnauthorizedAccessException) { }
+                return false;
+            }
+
+            private static bool Exists(ReadOnlySpan<char> path, int fileType, out Interop.ErrorInfo errorInfo)
```

Comment:
I would factor out the `FileSystem.Unix.cs`  Exists methods into their own file and share it (you can leave it in the same partial class). Converting to span for the existing FileExists will have the added benefit of allowing slicing off the directory separator instead of allocating a new string (when needed).